### PR TITLE
Search relays during mentions autocomplete

### PR DESCRIPTION
**Pull request by pelumy**
## Issues covered
#1560

## Description
This PR includes functionality for autocomplete to also search the query in the user's relay list . It contains the following changes:

- Refactors `AuthorListView` to use `SearchController`.

## How to test
1. Build the code.
2. Tap post button at the bottom of the screen.
3. Tap the @ button or type "@" into the textfield.
4. Type in the search bar.
5. Please check that Authors who weren't in Core Data before the AuthorListView opened are showing up in the list of results.

## Screenshots/Video

Before:

https://github.com/user-attachments/assets/a31797af-91f7-45e7-914a-d9fa8a4bb2dd

After:

https://github.com/user-attachments/assets/a48e355f-84a7-401e-bb73-d3d2ed9acd3d

### CODE REVIEW

File: CHANGELOG.md
Comment (joshuatbrown) [inline CHANGELOG.md:17]: 💯

File: Nos/Views/Components/Author/AuthorListView.swift
Comment (joshuatbrown) [inline Nos/Views/Components/Author/AuthorListView.swift:1]: This all looks so straightforward and simple. Nice!

**Review by joshuatbrown — COMMENTED** [review]
Everything looks good so far. I'm not sure I'm seeing results from relays, so maybe this isn't quite there yet.

**Comment by mplorentz** [comment]
👀

File: Nos/Views/Components/Author/AuthorListView.swift
Comment (mplorentz) [inline Nos/Views/Components/Author/AuthorListView.swift:42]: The `SearchController` has its own internal logic to fire off searches whenever the `query` changes. It has some advanced logic like debouncing and de-duplicating requests to make sure we don't send too many. I think we should get rid of this `onChange` and the `search(for query:)` function in this file. 

Maybe eventually we will need an `onChange(of: searchController.authorResults)` to do some custom sorting of the results and add participants in the thread. But for this PR I think we could even bind the `ForEach` directly to `searchController.authorResults` and get rid of the `filteredAuthors` array.

**Review by mplorentz — COMMENTED** [review]
I am seeing search results from the relays, but only after closing the dialog and searching for the same string a second time. [see video](https://github.com/user-attachments/assets/3c8ac00a-f35b-4c0e-9585-588918ee20ef)

(btw the [nostr.band trending page](https://nostr.band/trending/profiles/2024-10-08) is a nice place to find random folks to search for)

I think this is happening because we are only updating `filteredAuthors` in the `search(...)` function. In reality `search(...)` only starts the search, and the `SearchController` dynamically updates its `authorResults` variable after it has downloaded some results. Binding directly to the `authorResults`, or setting `filteredAuthors = searchController.authorResults` whenever `authorResults` updates would fix it I think.

**Comment by pelumy** [comment]
Thank you for the suggestions @mplorentz and thanks for pair programming with me on this @joshuatbrown

**Comment by pelumy** [comment]
After refactoring the `AuthorListView` based on @mplorentz detailed suggestion, it works perfectly now.
@joshuatbrown, also suggested making the keyboard button show "search" instead of "return"; and also searching the query when the user clicks this button.

This has also been added by changing the return button's label in the "SearchBar" view.

**Review by joshuatbrown — APPROVED** [review]
This looks great! One thing we discussed but I'm not sure how to handle: changing the SearchController's analytics tracking. It currently tracks "Discover Search Started", for example, when we're searching on the mentions screen. @mplorentz do you have a good sense of what we should track here? I'm assuming we want a new event along the lines of "Mentions Search Started", but I'm not sure if we also want an event comparable to the "Discover Search Displayed Author". I suppose so? Maybe call it "Mentions Search Selected Author"?

**Comment by pelumy** [comment]
> This looks great! One thing we discussed but I'm not sure how to handle: changing the SearchController's analytics tracking. It currently tracks "Discover Search Started", for example, when we're searching on the mentions screen. @mplorentz do you have a good sense of what we should track here? I'm assuming we want a new event along the lines of "Mentions Search Started", but I'm not sure if we also want an event comparable to the "Discover Search Displayed Author". I suppose so? Maybe call it "Mentions Search Selected Author"?

Aaah, thanks for bringing this up again. I totally forgot. I think `Selected` sounds like the user has selected an option. Looks like Linda created a new ticket for tracking mentions. But I feel we definitely need to differentiate the "Discover Search started" from the "Mentions search started" in this PR.

**Comment by mplorentz** [comment]
Oh good catch on the tracking. I agree that we shouldn't merge this if it's firing `Discover Search Started` from the mentions view, because that could mess up our production data.

Maybe we make a new enum called `SearchLocation` or something? And pass it into `SearchController.init` and then the Search controller can fire the right events? 

Duplicating the existing events but replacing "Discover" with "Mentions" sounds good to me.

**Comment by joshuatbrown** [comment]
👀

File: Nos/Controller/SearchController.swift
Comment (joshuatbrown) [inline Nos/Controller/SearchController.swift:32]: 💯 Love the name and the comments!

**Review by joshuatbrown — APPROVED** [review]
Works well and code looks great! Nice work!

**Pull request merged**
